lxd/storage/drivers/driver_btrfs.go

```diff
@@ -0,0 +1,1514 @@
+					} else if d.state.OS.BackingFS != "btrfs" {
+						return fmt.Errorf("Creation of btrfs subvolume requested but %q does not reside on btrfs filesystem", source)
+					}
+				}
+
+				err := btrfsSubvolumeCreate(source)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	var err error
+	var devUUID string
+	mountFlags, mountOptions := ResolveMountOptions(d.getMountOptions())
+	mountFlags |= d.remount
+
+	if isBlockDev {
+		devUUID, _ = shared.LookupUUIDByBlockDevPath(source)
+		// The symlink might not have been created even with the delay
+		// we granted it above. So try to call btrfs filesystem show and
+		// parse it out. (I __hate__ this!)
+		if devUUID == "" {
+			devUUID, err = btrfsLookupFsUUID(source)
+			if err != nil {
+				return err
+			}
+		}
+		d.config["source"] = devUUID
+
+		// If the symlink in /dev/disk/by-uuid hasn't been created yet
+		// aka we only detected it by parsing btrfs filesystem show, we
+		// cannot call StoragePoolMount() since it will try to do the
+		// reverse operation. So instead we shamelessly mount using the
+		// block device path at the time of pool creation.
+		err = tryMount(source, GetPoolMountPath(d.name), "btrfs", mountFlags, mountOptions)
+	} else {
+		_, err = d.Mount()
+	}
+	if err != nil {
+		return err
+	}
+
+	// Create default subvolumes.
+	subvolumes := []string{
+		filepath.Join(poolMntPoint, "containers"),
+		filepath.Join(poolMntPoint, "containers-snapshots"),
+		filepath.Join(poolMntPoint, "custom"),
+		filepath.Join(poolMntPoint, "custom-snapshots"),
+		filepath.Join(poolMntPoint, "images"),
+		filepath.Join(poolMntPoint, "virtual-machines"),
+		filepath.Join(poolMntPoint, "virtual-machines-snapshots"),
+	}
+
+	for _, subvol := range subvolumes {
+		err := btrfsSubvolumeCreate(subvol)
+		if err != nil {
+			return fmt.Errorf("Could not create btrfs subvolume: %s", subvol)
+		}
+	}
+
+	return nil
+}
+
+// Delete removes the storage pool from the storage device.
+func (d *btrfs) Delete(op *operations.Operation) error {
+	source := d.config["source"]
+
+	if source == "" {
+		return fmt.Errorf("no \"source\" property found for the storage pool")
+	}
+
+	if strings.HasPrefix(source, "/") {
+		source = shared.HostPath(d.config["source"])
+	}
+
+	poolMntPoint := GetPoolMountPath(d.name)
+
+	// Delete default subvolumes.
+	subvolumes := []string{
+		filepath.Join(poolMntPoint, "containers"),
+		filepath.Join(poolMntPoint, "containers-snapshots"),
+		filepath.Join(poolMntPoint, "custom"),
+		filepath.Join(poolMntPoint, "custom-snapshots"),
+		filepath.Join(poolMntPoint, "images"),
+		filepath.Join(poolMntPoint, "virtual-machines"),
+		filepath.Join(poolMntPoint, "virtual-machines-snapshots"),
+	}
+
+	for _, subvol := range subvolumes {
+		err := btrfsSubvolumesDelete(subvol)
+		if err != nil {
+			return fmt.Errorf("Could not delete btrfs subvolume: %s", subvol)
+		}
+	}
```

Comment:
I really don't like us hardcoding a list of potential subvolumes, can't we just iterate through all of them and get rid of them?